src/command/lsp/mod.rs

```diff
@@ -0,0 +1,346 @@
+            let diagnostic = Diagnostic::new_simple(Range::default(), message);
+            language_server
+                .publish_diagnostics(supergraph_yaml_url.clone(), vec![diagnostic])
+                .await;
+            return Err(e);
+        }
+    };
+
+    // Spawn a separate thread to handle composition and passing that data to the language server
+    tokio::spawn(async move {
+        while let Some(event) = stream.next().await {
+            match event {
+                CompositionEvent::Started => {
+                    // Even though it's hidden by library calls, this function emits a WorkDoneProgressBegin event,
+                    // which is paired with a WorkDoneProgressEnd event, sent by the `composition_did_update` function.
+                    // Any refactoring needs to ensure that we don't break this ordering, otherwise the LSP may well
+                    // cease to function in a useful way.
+                    language_server.composition_did_start().await;
+                }
+                CompositionEvent::Success(CompositionSuccess {
+                    supergraph_sdl,
+                    federation_version,
+                    hints,
+                }) => {
+                    debug!("Successfully composed with version {}", federation_version);
+                    // Clear any previous errors on `supergraph.yaml`
+                    language_server
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![])
+                        .await;
+                    language_server
+                        .composition_did_update(
+                            Some(supergraph_sdl),
+                            hints.into_iter().map(Into::into).collect(),
+                            None,
+                        )
+                        .await;
+                }
+                CompositionEvent::Error(CompositionError::Build {
+                    source: errors,
+                    federation_version,
+                }) => {
+                    debug!(
+                        ?errors,
+                        "Composition {federation_version} completed with errors"
+                    );
+                    // Clear any previous errors on `supergraph.yaml`
+                    language_server
+                        .publish_diagnostics(supergraph_yaml_url.clone(), vec![])
+                        .await;
+                    language_server
+                        .composition_did_update(
+                            None,
+                            errors.into_iter().map(Into::into).collect(),
+                            None,
+                        )
+                        .await
+                }
+                CompositionEvent::Error(err) => {
+                    debug!("Composition failed: {err}");
+                    let message = format!("Failed run composition: {err}",);
```

Comment:
```suggestion
                    let message = format!("Failed run composition: {err}");
```

WTB rustfmt which has an opinion on popular macros.